# Introduction to Deep Learning and Neural Network with Keras

**Final work**

**Ing. José Miguel Pereira Pontón**


# Instructions - Part A:
A. Build a baseline model (5 marks) 

Use the Keras library to build a neural network with the following:

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error  as the loss function.

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. you can use the train_test_splithelper function from Scikit-learn.

2. Train the model on the training data using 50 epochs.

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. you can use the mean_squared_error function from Scikit-learn.

4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

5. Report the mean and the standard deviation of the mean squared errors.

# Instructions - Part B:
6. Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.


# Preparing the environment

Import some libraries 

In [1]:
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.utils import to_categorical

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

import pandas as pd
import numpy as np

Read the data set

In [2]:
concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()



Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

Split dataset into inputs(predictors) and y (target)


In [3]:
# First: Split dataset into input (x) and output (y) variables
# Obtain the columns
concrete_data_columns = concrete_data.columns
# Selecting all columns except Strength
x = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] 
# Strength column
y = concrete_data['Strength']

In [4]:
# Lets do some review of information of predictors
x.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [5]:
# Lets do some review of information of target
y.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [6]:
# Finally normalizing the data by substracting the mean and dividing by the standard deviation
x_norm = (x-x.mean())/x.std()

In [7]:
# Number of predictors
n_cols = x.shape[1]

Using the model from Part A

In [8]:
# Defining a function for regression model
def regression_model():
    # create model
    model = Sequential()
    # First layer
    model.add(Dense(7, activation='relu',input_shape=(n_cols,)))
    # Hiden Layer
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    # Compile model
    model.compile(loss='mean_squared_error', optimizer='adam')
    return model

model = regression_model()

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_split helper function from Scikit-learn.

In [9]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x_norm, y, test_size=0.3, random_state=23)
# let's have a look at the shape of the predictors set
x_train.shape

(721, 8)

2. Train the model on the training data using 50 epochs.

In [10]:
model.fit(x_train, y_train, epochs=50)

Epoch 1/50
23/23 [==============================] - 0s 642us/step - loss: 1507.6321
Epoch 2/50
23/23 [==============================] - 0s 522us/step - loss: 1487.4158
Epoch 3/50
23/23 [==============================] - 0s 650us/step - loss: 1464.4225
Epoch 4/50
23/23 [==============================] - 0s 648us/step - loss: 1436.5731
Epoch 5/50
23/23 [==============================] - 0s 694us/step - loss: 1401.3430
Epoch 6/50
23/23 [==============================] - 0s 607us/step - loss: 1357.2477
Epoch 7/50
23/23 [==============================] - 0s 520us/step - loss: 1303.3502
Epoch 8/50
23/23 [==============================] - 0s 622us/step - loss: 1238.3943
Epoch 9/50
23/23 [==============================] - 0s 780us/step - loss: 1162.1090
Epoch 10/50
23/23 [==============================] - 0s 692us/step - loss: 1076.3389
Epoch 11/50
23/23 [==============================] - 0s 650us/step - loss: 982.4610
Epoch 12/50
23/23 [==============================] - 0s 650us/step - loss: 

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. you can use the mean_squared_error function from Scikit-learn.


In [11]:
y_pred = model.predict(x_test)
from sklearn.metrics import mean_squared_error
print('Mean squared error on test data is %.3f' % (mean_squared_error(y_test, y_pred)))

Mean squared error on test data is 159.145


4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.
5. Report the mean and the standard deviation of the mean squared errors.

In [12]:
mse = []
for i in range(50):
    model = regression_model()
    model.fit(x_train, y_train, epochs=50)
    y_pred = model.predict(x_test)
    mse.append(mean_squared_error(y_test, y_pred))

Epoch 1/50
23/23 [==============================] - 0s 1ms/step - loss: 1545.4241
Epoch 2/50
23/23 [==============================] - 0s 1ms/step - loss: 1526.1292
Epoch 3/50
23/23 [==============================] - 0s 1ms/step - loss: 1505.8357
Epoch 4/50
23/23 [==============================] - 0s 1ms/step - loss: 1482.7195
Epoch 5/50
23/23 [==============================] - 0s 1ms/step - loss: 1454.1011
Epoch 6/50
23/23 [==============================] - 0s 2ms/step - loss: 1418.7914
Epoch 7/50
23/23 [==============================] - 0s 2ms/step - loss: 1372.7549
Epoch 8/50
23/23 [==============================] - 0s 2ms/step - loss: 1315.7203
Epoch 9/50
23/23 [==============================] - 0s 2ms/step - loss: 1245.1249
Epoch 10/50
23/23 [==============================] - 0s 1ms/step - loss: 1162.7612
Epoch 11/50
23/23 [==============================] - 0s 2ms/step - loss: 1071.4349
Epoch 12/50
23/23 [==============================] - 0s 2ms/step - loss: 974.1504
Epoch 13/50
23

23/23 [==============================] - 0s 1ms/step - loss: 146.9297
Epoch 1/50
23/23 [==============================] - 0s 997us/step - loss: 1456.8235
Epoch 2/50
23/23 [==============================] - 0s 912us/step - loss: 1422.3955
Epoch 3/50
23/23 [==============================] - 0s 991us/step - loss: 1381.5819
Epoch 4/50
23/23 [==============================] - 0s 1ms/step - loss: 1332.0466
Epoch 5/50
23/23 [==============================] - 0s 1ms/step - loss: 1272.9312
Epoch 6/50
23/23 [==============================] - 0s 1ms/step - loss: 1200.1509
Epoch 7/50
23/23 [==============================] - 0s 1ms/step - loss: 1114.5349
Epoch 8/50
23/23 [==============================] - 0s 1ms/step - loss: 1016.1945
Epoch 9/50
23/23 [==============================] - 0s 1ms/step - loss: 908.0851
Epoch 10/50
23/23 [==============================] - 0s 1ms/step - loss: 793.9689
Epoch 11/50
23/23 [==============================] - 0s 1ms/step - loss: 679.4531
Epoch 12/50
23/23 [====

23/23 [==============================] - 0s 737us/step - loss: 160.7452
Epoch 50/50
23/23 [==============================] - 0s 650us/step - loss: 159.5511
Epoch 1/50
23/23 [==============================] - 0s 1ms/step - loss: 1515.6344
Epoch 2/50
23/23 [==============================] - 0s 996us/step - loss: 1490.9052
Epoch 3/50
23/23 [==============================] - 0s 954us/step - loss: 1463.2231
Epoch 4/50
23/23 [==============================] - 0s 997us/step - loss: 1430.8704
Epoch 5/50
23/23 [==============================] - 0s 997us/step - loss: 1391.8207
Epoch 6/50
23/23 [==============================] - 0s 877us/step - loss: 1345.6566
Epoch 7/50
23/23 [==============================] - 0s 1ms/step - loss: 1289.6117
Epoch 8/50
23/23 [==============================] - 0s 1ms/step - loss: 1223.9547
Epoch 9/50
23/23 [==============================] - 0s 911us/step - loss: 1148.4795
Epoch 10/50
23/23 [==============================] - 0s 826us/step - loss: 1060.2555
Epoch 11/

23/23 [==============================] - 0s 1ms/step - loss: 151.1058
Epoch 49/50
23/23 [==============================] - 0s 1ms/step - loss: 150.2499
Epoch 50/50
23/23 [==============================] - 0s 1ms/step - loss: 149.4586
Epoch 1/50
23/23 [==============================] - 0s 1ms/step - loss: 1532.4806
Epoch 2/50
23/23 [==============================] - 0s 1ms/step - loss: 1513.7365
Epoch 3/50
23/23 [==============================] - 0s 1ms/step - loss: 1493.6056
Epoch 4/50
23/23 [==============================] - 0s 1ms/step - loss: 1469.4161
Epoch 5/50
23/23 [==============================] - 0s 1ms/step - loss: 1440.2494
Epoch 6/50
23/23 [==============================] - 0s 1ms/step - loss: 1404.1438
Epoch 7/50
23/23 [==============================] - 0s 1ms/step - loss: 1359.0087
Epoch 8/50
23/23 [==============================] - 0s 1ms/step - loss: 1302.3833
Epoch 9/50
23/23 [==============================] - 0s 1ms/step - loss: 1235.5750
Epoch 10/50
23/23 [=========

23/23 [==============================] - 0s 997us/step - loss: 149.4945
Epoch 48/50
23/23 [==============================] - 0s 1ms/step - loss: 148.1087
Epoch 49/50
23/23 [==============================] - 0s 1ms/step - loss: 146.9841
Epoch 50/50
23/23 [==============================] - 0s 1ms/step - loss: 145.6699
Epoch 1/50
23/23 [==============================] - 0s 1ms/step - loss: 1483.5106
Epoch 2/50
23/23 [==============================] - 0s 1ms/step - loss: 1457.3254
Epoch 3/50
23/23 [==============================] - 0s 1ms/step - loss: 1426.9141
Epoch 4/50
23/23 [==============================] - 0s 1ms/step - loss: 1390.6207
Epoch 5/50
23/23 [==============================] - 0s 1ms/step - loss: 1346.1929
Epoch 6/50
23/23 [==============================] - 0s 1ms/step - loss: 1291.0226
Epoch 7/50
23/23 [==============================] - 0s 1ms/step - loss: 1226.7047
Epoch 8/50
23/23 [==============================] - 0s 1ms/step - loss: 1153.4521
Epoch 9/50
23/23 [========

23/23 [==============================] - 0s 1ms/step - loss: 158.5514
Epoch 47/50
23/23 [==============================] - 0s 1ms/step - loss: 156.8967
Epoch 48/50
23/23 [==============================] - 0s 1ms/step - loss: 155.4770
Epoch 49/50
23/23 [==============================] - 0s 1ms/step - loss: 153.9635
Epoch 50/50
23/23 [==============================] - 0s 1ms/step - loss: 152.5739
Epoch 1/50
23/23 [==============================] - 0s 952us/step - loss: 1591.7888
Epoch 2/50
23/23 [==============================] - 0s 863us/step - loss: 1573.5459
Epoch 3/50
23/23 [==============================] - 0s 716us/step - loss: 1559.5768
Epoch 4/50
23/23 [==============================] - 0s 822us/step - loss: 1548.6544
Epoch 5/50
23/23 [==============================] - 0s 911us/step - loss: 1539.2439
Epoch 6/50
23/23 [==============================] - 0s 804us/step - loss: 1530.2545
Epoch 7/50
23/23 [==============================] - 0s 709us/step - loss: 1520.6320
Epoch 8/50
23/

23/23 [==============================] - 0s 651us/step - loss: 163.9745
Epoch 46/50
23/23 [==============================] - 0s 652us/step - loss: 162.0420
Epoch 47/50
23/23 [==============================] - 0s 651us/step - loss: 160.3025
Epoch 48/50
23/23 [==============================] - 0s 694us/step - loss: 158.6066
Epoch 49/50
23/23 [==============================] - 0s 648us/step - loss: 156.9375
Epoch 50/50
23/23 [==============================] - 0s 720us/step - loss: 155.3639
Epoch 1/50
23/23 [==============================] - 0s 995us/step - loss: 1511.6482
Epoch 2/50
23/23 [==============================] - 0s 1ms/step - loss: 1488.4811
Epoch 3/50
23/23 [==============================] - 0s 1ms/step - loss: 1462.8911
Epoch 4/50
23/23 [==============================] - 0s 1ms/step - loss: 1433.3059
Epoch 5/50
23/23 [==============================] - 0s 1ms/step - loss: 1396.7758
Epoch 6/50
23/23 [==============================] - 0s 1ms/step - loss: 1351.7483
Epoch 7/50
23/

23/23 [==============================] - 0s 1ms/step - loss: 169.2358
Epoch 45/50
23/23 [==============================] - 0s 1ms/step - loss: 167.6328
Epoch 46/50
23/23 [==============================] - 0s 1ms/step - loss: 166.1807
Epoch 47/50
23/23 [==============================] - 0s 1ms/step - loss: 164.8476
Epoch 48/50
23/23 [==============================] - 0s 1ms/step - loss: 163.6066
Epoch 49/50
23/23 [==============================] - 0s 1ms/step - loss: 162.3816
Epoch 50/50
23/23 [==============================] - 0s 1ms/step - loss: 161.3020
Epoch 1/50
23/23 [==============================] - 0s 2ms/step - loss: 1520.0184
Epoch 2/50
23/23 [==============================] - 0s 2ms/step - loss: 1500.4532
Epoch 3/50
23/23 [==============================] - 0s 2ms/step - loss: 1478.5812
Epoch 4/50
23/23 [==============================] - 0s 2ms/step - loss: 1450.1843
Epoch 5/50
23/23 [==============================] - 0s 2ms/step - loss: 1413.8040
Epoch 6/50
23/23 [==========

23/23 [==============================] - 0s 1ms/step - loss: 142.8548
Epoch 45/50
23/23 [==============================] - 0s 1ms/step - loss: 141.4023
Epoch 46/50
23/23 [==============================] - 0s 1ms/step - loss: 139.8316
Epoch 47/50
23/23 [==============================] - 0s 1ms/step - loss: 138.3378
Epoch 48/50
23/23 [==============================] - 0s 1ms/step - loss: 136.9013
Epoch 49/50
23/23 [==============================] - 0s 1ms/step - loss: 135.5549
Epoch 50/50
23/23 [==============================] - 0s 1ms/step - loss: 134.3256
Epoch 1/50
23/23 [==============================] - 0s 954us/step - loss: 1569.2295
Epoch 2/50
23/23 [==============================] - 0s 1ms/step - loss: 1549.8271
Epoch 3/50
23/23 [==============================] - 0s 1ms/step - loss: 1532.9459
Epoch 4/50
23/23 [==============================] - 0s 1ms/step - loss: 1517.3660
Epoch 5/50
23/23 [==============================] - 0s 1ms/step - loss: 1500.9200
Epoch 6/50
23/23 [========

23/23 [==============================] - 0s 1ms/step - loss: 158.5089
Epoch 44/50
23/23 [==============================] - 0s 1ms/step - loss: 157.3110
Epoch 45/50
23/23 [==============================] - 0s 1ms/step - loss: 156.1057
Epoch 46/50
23/23 [==============================] - 0s 1ms/step - loss: 154.8039
Epoch 47/50
23/23 [==============================] - 0s 1ms/step - loss: 153.8290
Epoch 48/50
23/23 [==============================] - 0s 1ms/step - loss: 152.6996
Epoch 49/50
23/23 [==============================] - 0s 1ms/step - loss: 151.7232
Epoch 50/50
23/23 [==============================] - 0s 1ms/step - loss: 150.8052
Epoch 1/50
23/23 [==============================] - 0s 1ms/step - loss: 1564.5946
Epoch 2/50
23/23 [==============================] - 0s 1ms/step - loss: 1550.2322
Epoch 3/50
23/23 [==============================] - 0s 1ms/step - loss: 1537.8563
Epoch 4/50
23/23 [==============================] - 0s 996us/step - loss: 1525.6165
Epoch 5/50
23/23 [========

23/23 [==============================] - 0s 672us/step - loss: 153.0416
Epoch 42/50
23/23 [==============================] - 0s 608us/step - loss: 151.8525
Epoch 43/50
23/23 [==============================] - 0s 658us/step - loss: 150.4666
Epoch 44/50
23/23 [==============================] - 0s 648us/step - loss: 149.2000
Epoch 45/50
23/23 [==============================] - 0s 522us/step - loss: 148.1441
Epoch 46/50
23/23 [==============================] - 0s 520us/step - loss: 146.6870
Epoch 47/50
23/23 [==============================] - 0s 739us/step - loss: 145.5667
Epoch 48/50
23/23 [==============================] - 0s 694us/step - loss: 144.5421
Epoch 49/50
23/23 [==============================] - 0s 952us/step - loss: 143.7314
Epoch 50/50
23/23 [==============================] - 0s 1ms/step - loss: 142.5721
Epoch 1/50
23/23 [==============================] - 0s 2ms/step - loss: 1515.3683
Epoch 2/50
23/23 [==============================] - 0s 2ms/step - loss: 1493.5919
Epoch 3/50

23/23 [==============================] - 0s 1ms/step - loss: 176.4417
Epoch 41/50
23/23 [==============================] - 0s 1ms/step - loss: 173.5405
Epoch 42/50
23/23 [==============================] - 0s 1ms/step - loss: 170.6682
Epoch 43/50
23/23 [==============================] - 0s 1ms/step - loss: 168.3010
Epoch 44/50
23/23 [==============================] - 0s 1ms/step - loss: 165.6193
Epoch 45/50
23/23 [==============================] - 0s 1ms/step - loss: 163.3561
Epoch 46/50
23/23 [==============================] - 0s 1ms/step - loss: 161.2525
Epoch 47/50
23/23 [==============================] - 0s 1ms/step - loss: 159.2274
Epoch 48/50
23/23 [==============================] - 0s 1ms/step - loss: 157.3447
Epoch 49/50
23/23 [==============================] - 0s 1ms/step - loss: 155.1537
Epoch 50/50
23/23 [==============================] - 0s 1ms/step - loss: 153.3494
Epoch 1/50
23/23 [==============================] - 0s 1ms/step - loss: 1537.7343
Epoch 2/50
23/23 [==========

23/23 [==============================] - 0s 1ms/step - loss: 174.8120
Epoch 40/50
23/23 [==============================] - 0s 1ms/step - loss: 173.0271
Epoch 41/50
23/23 [==============================] - 0s 1ms/step - loss: 171.1360
Epoch 42/50
23/23 [==============================] - 0s 1ms/step - loss: 169.4643
Epoch 43/50
23/23 [==============================] - 0s 1ms/step - loss: 167.6571
Epoch 44/50
23/23 [==============================] - 0s 1ms/step - loss: 166.0051
Epoch 45/50
23/23 [==============================] - 0s 1ms/step - loss: 164.3083
Epoch 46/50
23/23 [==============================] - 0s 1ms/step - loss: 162.6058
Epoch 47/50
23/23 [==============================] - 0s 1ms/step - loss: 161.0381
Epoch 48/50
23/23 [==============================] - 0s 1ms/step - loss: 159.6206
Epoch 49/50
23/23 [==============================] - 0s 1ms/step - loss: 158.1049
Epoch 50/50
23/23 [==============================] - 0s 1ms/step - loss: 156.7856
Epoch 1/50
23/23 [==========

23/23 [==============================] - 0s 780us/step - loss: 176.3714
Epoch 39/50
23/23 [==============================] - 0s 788us/step - loss: 173.9946
Epoch 40/50
23/23 [==============================] - 0s 607us/step - loss: 171.6082
Epoch 41/50
23/23 [==============================] - 0s 613us/step - loss: 169.7254
Epoch 42/50
23/23 [==============================] - 0s 826us/step - loss: 167.3562
Epoch 43/50
23/23 [==============================] - 0s 991us/step - loss: 165.3879
Epoch 44/50
23/23 [==============================] - 0s 694us/step - loss: 163.1592
Epoch 45/50
23/23 [==============================] - 0s 1ms/step - loss: 161.2609
Epoch 46/50
23/23 [==============================] - 0s 997us/step - loss: 159.4077
Epoch 47/50
23/23 [==============================] - 0s 672us/step - loss: 157.6894
Epoch 48/50
23/23 [==============================] - 0s 737us/step - loss: 156.0983
Epoch 49/50
23/23 [==============================] - 0s 694us/step - loss: 154.5936
Epoch 

23/23 [==============================] - 0s 1ms/step - loss: 185.1208
Epoch 37/50
23/23 [==============================] - 0s 1ms/step - loss: 183.3987
Epoch 38/50
23/23 [==============================] - 0s 1ms/step - loss: 181.5128
Epoch 39/50
23/23 [==============================] - 0s 1ms/step - loss: 180.0804
Epoch 40/50
23/23 [==============================] - 0s 1ms/step - loss: 177.9669
Epoch 41/50
23/23 [==============================] - 0s 1ms/step - loss: 176.3405
Epoch 42/50
23/23 [==============================] - 0s 1ms/step - loss: 174.8657
Epoch 43/50
23/23 [==============================] - 0s 1ms/step - loss: 173.1228
Epoch 44/50
23/23 [==============================] - 0s 1ms/step - loss: 171.7042
Epoch 45/50
23/23 [==============================] - 0s 1ms/step - loss: 170.1503
Epoch 46/50
23/23 [==============================] - 0s 1ms/step - loss: 168.6277
Epoch 47/50
23/23 [==============================] - 0s 1ms/step - loss: 167.2682
Epoch 48/50
23/23 [=========

23/23 [==============================] - 0s 997us/step - loss: 180.6564
Epoch 36/50
23/23 [==============================] - 0s 999us/step - loss: 178.4229
Epoch 37/50
23/23 [==============================] - 0s 1000us/step - loss: 176.6930
Epoch 38/50
23/23 [==============================] - 0s 1ms/step - loss: 174.8602
Epoch 39/50
23/23 [==============================] - 0s 1ms/step - loss: 173.1842
Epoch 40/50
23/23 [==============================] - 0s 997us/step - loss: 171.7679
Epoch 41/50
23/23 [==============================] - 0s 1ms/step - loss: 170.1948
Epoch 42/50
23/23 [==============================] - 0s 1ms/step - loss: 168.7874
Epoch 43/50
23/23 [==============================] - 0s 1ms/step - loss: 167.3692
Epoch 44/50
23/23 [==============================] - 0s 1ms/step - loss: 165.9955
Epoch 45/50
23/23 [==============================] - 0s 1ms/step - loss: 164.6636
Epoch 46/50
23/23 [==============================] - 0s 1ms/step - loss: 163.3463
Epoch 47/50
23/23 [

23/23 [==============================] - 0s 1ms/step - loss: 192.7363
Epoch 35/50
23/23 [==============================] - 0s 1ms/step - loss: 189.9968
Epoch 36/50
23/23 [==============================] - 0s 1ms/step - loss: 187.6568
Epoch 37/50
23/23 [==============================] - 0s 1ms/step - loss: 185.4093
Epoch 38/50
23/23 [==============================] - 0s 1ms/step - loss: 183.3851
Epoch 39/50
23/23 [==============================] - 0s 1ms/step - loss: 181.2525
Epoch 40/50
23/23 [==============================] - 0s 1ms/step - loss: 179.4151
Epoch 41/50
23/23 [==============================] - 0s 1ms/step - loss: 177.7173
Epoch 42/50
23/23 [==============================] - 0s 1ms/step - loss: 175.9046
Epoch 43/50
23/23 [==============================] - 0s 1ms/step - loss: 174.3389
Epoch 44/50
23/23 [==============================] - 0s 1ms/step - loss: 172.8082
Epoch 45/50
23/23 [==============================] - 0s 997us/step - loss: 171.1997
Epoch 46/50
23/23 [=======

23/23 [==============================] - 0s 956us/step - loss: 205.6070
Epoch 33/50
23/23 [==============================] - 0s 1ms/step - loss: 201.2709
Epoch 34/50
23/23 [==============================] - 0s 1ms/step - loss: 197.2848
Epoch 35/50
23/23 [==============================] - 0s 1ms/step - loss: 193.6109
Epoch 36/50
23/23 [==============================] - 0s 1ms/step - loss: 190.2837
Epoch 37/50
23/23 [==============================] - 0s 1ms/step - loss: 186.8184
Epoch 38/50
23/23 [==============================] - 0s 1ms/step - loss: 183.7755
Epoch 39/50
23/23 [==============================] - 0s 1ms/step - loss: 180.8145
Epoch 40/50
23/23 [==============================] - 0s 1ms/step - loss: 178.3256
Epoch 41/50
23/23 [==============================] - 0s 1ms/step - loss: 175.7428
Epoch 42/50
23/23 [==============================] - 0s 1ms/step - loss: 173.1387
Epoch 43/50
23/23 [==============================] - 0s 1ms/step - loss: 170.9655
Epoch 44/50
23/23 [=======

23/23 [==============================] - 0s 1ms/step - loss: 183.7222
Epoch 32/50
23/23 [==============================] - 0s 1ms/step - loss: 181.0276
Epoch 33/50
23/23 [==============================] - 0s 1ms/step - loss: 178.6143
Epoch 34/50
23/23 [==============================] - 0s 1ms/step - loss: 176.5295
Epoch 35/50
23/23 [==============================] - 0s 1ms/step - loss: 174.3522
Epoch 36/50
23/23 [==============================] - 0s 1ms/step - loss: 172.3387
Epoch 37/50
23/23 [==============================] - 0s 1ms/step - loss: 170.6074
Epoch 38/50
23/23 [==============================] - 0s 999us/step - loss: 168.8464
Epoch 39/50
23/23 [==============================] - 0s 1ms/step - loss: 167.3117
Epoch 40/50
23/23 [==============================] - 0s 1ms/step - loss: 165.5139
Epoch 41/50
23/23 [==============================] - 0s 1ms/step - loss: 164.0942
Epoch 42/50
23/23 [==============================] - 0s 1ms/step - loss: 162.3073
Epoch 43/50
23/23 [=======

23/23 [==============================] - 0s 984us/step - loss: 174.4808
Epoch 31/50
23/23 [==============================] - 0s 1ms/step - loss: 171.7602
Epoch 32/50
23/23 [==============================] - 0s 912us/step - loss: 169.3640
Epoch 33/50
23/23 [==============================] - 0s 1ms/step - loss: 166.9713
Epoch 34/50
23/23 [==============================] - 0s 1ms/step - loss: 164.9916
Epoch 35/50
23/23 [==============================] - 0s 1ms/step - loss: 163.3747
Epoch 36/50
23/23 [==============================] - 0s 1ms/step - loss: 161.4585
Epoch 37/50
23/23 [==============================] - 0s 1ms/step - loss: 159.8273
Epoch 38/50
23/23 [==============================] - 0s 1ms/step - loss: 158.4846
Epoch 39/50
23/23 [==============================] - 0s 1ms/step - loss: 156.8947
Epoch 40/50
23/23 [==============================] - 0s 1ms/step - loss: 155.6558
Epoch 41/50
23/23 [==============================] - 0s 1ms/step - loss: 154.2040
Epoch 42/50
23/23 [=====

23/23 [==============================] - 0s 1ms/step - loss: 203.1817
Epoch 29/50
23/23 [==============================] - 0s 1ms/step - loss: 198.0544
Epoch 30/50
23/23 [==============================] - 0s 1ms/step - loss: 194.1436
Epoch 31/50
23/23 [==============================] - 0s 1ms/step - loss: 190.7659
Epoch 32/50
23/23 [==============================] - 0s 1ms/step - loss: 187.6639
Epoch 33/50
23/23 [==============================] - 0s 1ms/step - loss: 184.9132
Epoch 34/50
23/23 [==============================] - 0s 1ms/step - loss: 182.2126
Epoch 35/50
23/23 [==============================] - 0s 1ms/step - loss: 179.7361
Epoch 36/50
23/23 [==============================] - 0s 2ms/step - loss: 177.3575
Epoch 37/50
23/23 [==============================] - 0s 1ms/step - loss: 175.1590
Epoch 38/50
23/23 [==============================] - 0s 1ms/step - loss: 172.8175
Epoch 39/50
23/23 [==============================] - 0s 1ms/step - loss: 170.8149
Epoch 40/50
23/23 [=========

23/23 [==============================] - 0s 1ms/step - loss: 187.8753
Epoch 28/50
23/23 [==============================] - 0s 1ms/step - loss: 182.7429
Epoch 29/50
23/23 [==============================] - 0s 1ms/step - loss: 178.4230
Epoch 30/50
23/23 [==============================] - 0s 1ms/step - loss: 175.1863
Epoch 31/50
23/23 [==============================] - 0s 1ms/step - loss: 172.0732
Epoch 32/50
23/23 [==============================] - ETA: 0s - loss: 228.384 - 0s 615us/step - loss: 169.6146
Epoch 33/50
23/23 [==============================] - 0s 564us/step - loss: 167.0922
Epoch 34/50
23/23 [==============================] - 0s 566us/step - loss: 165.3095
Epoch 35/50
23/23 [==============================] - 0s 606us/step - loss: 163.1848
Epoch 36/50
23/23 [==============================] - 0s 604us/step - loss: 161.5386
Epoch 37/50
23/23 [==============================] - 0s 589us/step - loss: 159.9924
Epoch 38/50
23/23 [==============================] - 0s 563us/step - los

23/23 [==============================] - 0s 1ms/step - loss: 212.8075
Epoch 26/50
23/23 [==============================] - 0s 1ms/step - loss: 205.1450
Epoch 27/50
23/23 [==============================] - 0s 1ms/step - loss: 199.3105
Epoch 28/50
23/23 [==============================] - 0s 1ms/step - loss: 194.4613
Epoch 29/50
23/23 [==============================] - 0s 1ms/step - loss: 189.9762
Epoch 30/50
23/23 [==============================] - 0s 1ms/step - loss: 186.2494
Epoch 31/50
23/23 [==============================] - 0s 1ms/step - loss: 182.9937
Epoch 32/50
23/23 [==============================] - 0s 1ms/step - loss: 180.2052
Epoch 33/50
23/23 [==============================] - 0s 1ms/step - loss: 177.6825
Epoch 34/50
23/23 [==============================] - 0s 1ms/step - loss: 175.4953
Epoch 35/50
23/23 [==============================] - 0s 1ms/step - loss: 173.1980
Epoch 36/50
23/23 [==============================] - 0s 1ms/step - loss: 171.2847
Epoch 37/50
23/23 [=========

23/23 [==============================] - 0s 1ms/step - loss: 609.6483
Epoch 25/50
23/23 [==============================] - 0s 1ms/step - loss: 551.5190
Epoch 26/50
23/23 [==============================] - 0s 1ms/step - loss: 499.1257
Epoch 27/50
23/23 [==============================] - 0s 1ms/step - loss: 450.7954
Epoch 28/50
23/23 [==============================] - 0s 1ms/step - loss: 407.4205
Epoch 29/50
23/23 [==============================] - 0s 1ms/step - loss: 368.3482
Epoch 30/50
23/23 [==============================] - 0s 1ms/step - loss: 334.0577
Epoch 31/50
23/23 [==============================] - 0s 1ms/step - loss: 303.9330
Epoch 32/50
23/23 [==============================] - 0s 1ms/step - loss: 278.4629
Epoch 33/50
23/23 [==============================] - 0s 1ms/step - loss: 258.2549
Epoch 34/50
23/23 [==============================] - 0s 1ms/step - loss: 242.4260
Epoch 35/50
23/23 [==============================] - 0s 1ms/step - loss: 229.5688
Epoch 36/50
23/23 [=========

23/23 [==============================] - 0s 1ms/step - loss: 203.9189
Epoch 24/50
23/23 [==============================] - 0s 1ms/step - loss: 196.7137
Epoch 25/50
23/23 [==============================] - 0s 1ms/step - loss: 190.8597
Epoch 26/50
23/23 [==============================] - 0s 1ms/step - loss: 186.1113
Epoch 27/50
23/23 [==============================] - 0s 1ms/step - loss: 182.0378
Epoch 28/50
23/23 [==============================] - 0s 1ms/step - loss: 178.6458
Epoch 29/50
23/23 [==============================] - 0s 1ms/step - loss: 175.5211
Epoch 30/50
23/23 [==============================] - 0s 1ms/step - loss: 172.6764
Epoch 31/50
23/23 [==============================] - 0s 1ms/step - loss: 170.0555
Epoch 32/50
23/23 [==============================] - 0s 1ms/step - loss: 167.9663
Epoch 33/50
23/23 [==============================] - 0s 1ms/step - loss: 165.5185
Epoch 34/50
23/23 [==============================] - 0s 1ms/step - loss: 163.4499
Epoch 35/50
23/23 [=========

23/23 [==============================] - 0s 1ms/step - loss: 324.7026
Epoch 23/50
23/23 [==============================] - 0s 1ms/step - loss: 294.7197
Epoch 24/50
23/23 [==============================] - 0s 1ms/step - loss: 270.6887
Epoch 25/50
23/23 [==============================] - 0s 1ms/step - loss: 252.0146
Epoch 26/50
23/23 [==============================] - 0s 1ms/step - loss: 236.0283
Epoch 27/50
23/23 [==============================] - 0s 737us/step - loss: 223.0588
Epoch 28/50
23/23 [==============================] - 0s 824us/step - loss: 212.8404
Epoch 29/50
23/23 [==============================] - 0s 865us/step - loss: 204.6490
Epoch 30/50
23/23 [==============================] - 0s 693us/step - loss: 197.8031
Epoch 31/50
23/23 [==============================] - 0s 694us/step - loss: 192.1869
Epoch 32/50
23/23 [==============================] - 0s 691us/step - loss: 187.2416
Epoch 33/50
23/23 [==============================] - 0s 639us/step - loss: 183.0139
Epoch 34/50
23

5. Report the mean and the standard deviation of the mean squared errors.

In [13]:
mean = np.mean(mse)
mean

151.0774193638029

In [14]:
std = np.std(mse)
std

10.43360202711478

In [15]:
print("Mean squared errors for 50 regression models: mean = %.3f, std dev = %.3f" %(mean, std))

Mean squared errors for 50 regression models: mean = 151.077, std dev = 10.434


How does the mean of the mean squared errors compare to that from Step A?
In Step A, I record a MSE of 176.835 and in Part B I record a MSE of 159.145, which is significantly better.

And if we review the MSE of the 50 regression model:

|--- |Part A | Part B|
| --- | ---- | --- |
|**MSE**|203.422 | 151.077|
| **STD** | 206.667 | 10.433|

We can determine that the result of the regression is improved when we normalize the data in from of when we use the data in the normal way.

**End of Part B.**